### ISSUE TITLE
Compatibility with Cookiebot

### ISSUE THREAD

**Issue by juricazuanovic** [wp-media/wp-rocket#4062]
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version
- Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
In order for [Cookiebot](https://www.cookiebot.com/en/) to work with WP Rocket, it is required to exclude the following from Minify JavaScript files/Combine JavaScript files:

`consent.cookiebot.com`
`consentcdn.cookiebot.com`

**Expected behavior**
This should be excluded by default so that WP Rocket works with Cookiebot out of the box.

**Additional context**
Ticket: https://secure.helpscout.net/conversation/1527176652/267585/

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

**Comment by gevcen** [comment]
Since Combine JS is expected to be deprecated https://github.com/wp-media/wp-rocket/issues/4023 and depending on the priority of this issue here I suppose it's relevant to only exclude these files from Minify JS only?

**Comment by remyperona** [comment]
### Scope a solution ✅ 
Add `consent.cookiebot.com` and `consentcdn.cookiebot.com` as patterns to the defaults exclusion array in `AbstractJSOptimization::get_excluded_external_file_path()`

### Estimate the effort ✅
Effort `[XS]`

**Comment by Mai-Saad** [comment]
@juricazuanovic @piotrbak Please find the notes below from the exploratory test:

1- Cookiebot is working fine on trunk without any exclusions (by default Cookiebot is excluded from Minify/Combine JS) , do we still need to add auto exclusion here?
i.e this script is not minified nor combined  
 `<script type="rocketlazyloadscript" data-rocket-type="text/javascript"
  |id="Cookiebot"
 src="https://consent.cookiebot.com/uc.js"
data-cbid="8##########################"</script>
`

2-  Complianz plugin is working with  Minify/Combine JS  but with certain test template we have to add exclusions to make it work. Do we need to handle anything for the test template in a separate GH?

For Minify JS

`complianz.min.js`
`cookieconsent.min.js`  

 While for Combine JS, we need to exclude  `jquery`  too from js files

**Comment by engahmeds3ed** [comment]
I can see that Cookiebot plugin uses our filter to add those exclusions here: 

https://github.com/CybotAS/CookiebotWP/blob/master/cookiebot.php#L195-L198

https://github.com/CybotAS/CookiebotWP/blob/master/cookiebot.php#L2224-L2234

I don't think that we need to add it from our side too, I'd suggest closing this PR.

**Comment by juricazuanovic** [comment]
Since this is now excluded on Cookiebot side, we can close this.

**Comment by piotrbak** [comment]
Closing it, since the issue is fixed on their end.

**Comment by vmanthos** [comment]
Despite Cookiebot adding the exclusions on their side, these aren't really applied.

This is because the following conditional isn't met:
https://github.com/CybotAS/CookiebotWP/blob/befb56c106f642d688794ac3e5fc2e3a06be9f19/src/lib/WP_Rocket_Helper.php#L8-L10

Changing it to `if ( function_exists( 'rocket_clean_domain' ) )` doesn't change anything. 

@piotrbak We need to decide whether we'll reach out to them and ask them to look into this or handle it on our side.

Props to @camilamadronero and @webtrainingwheels for reporting the issue.

**Ticket:** https://secure.helpscout.net/conversation/1768350384/321848?folderId=2135277
**Slack convo:** https://wp-media.slack.com/archives/C43T1AYMQ/p1643647279909959

**Comment by piotrbak** [comment]
@vmanthos we should handle that on our side. Depending on others doesn't seem to be the best way out.

**Comment by vmanthos** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1793965046/327462/

Excluding `consent.cookiebot.com/uc.js` from JavaScript minification resolved the issue.

**Comment by DahmaniAdame** [comment]
Related - https://secure.helpscout.net/conversation/1782506457/325057/

Excluding `consent.cookiebot.com/uc.js` from minification fixes the issue. 

As a precaution, I added the same exclusions that CookieBot is trying to do here: [/src/lib/WP_Rocket_Helper.php](https://github.com/CybotAS/CookiebotWP/blob/4c107e8870564b0949d2531c627ccb563ed9f708/src/lib/WP_Rocket_Helper.php)

```
consent.cookiebot.com
consentcdn.cookiebot.com
```

**Comment by webtrainingwheels** [comment]
related https://secure.helpscout.net/conversation/1808246899/330092?folderId=377611

**Comment by DahmaniAdame** [comment]
Related - https://secure.helpscout.net/conversation/1932295986/350802/

**Comment by NataliaDrause** [comment]
I found that `consent.cookiebot.com` is now automatically excluded from Minify JavaScript option, but there is an issue caused by the Delay JavaScript Execution.

Tested on WP Rocket (3.11.4) and [Cookiebot](https://wordpress.org/plugins/cookiebot/) (4.1.0).

Steps to reproduce (reproduced on my test site):

1. Install and activate **[Cookiebot | GDPR/CCPA Compliant Cookie Consent and Control](https://wordpress.org/plugins/cookiebot/)** plugin.
2. You will need to create a free Cookiebot account to receive an ID.
3. Add the ID in the settings. Cookiebot settings for reference: https://i.imgur.com/oxJvsOp.png
4. Enable JavaScript optimization options: Minify JavaScript, Load JavaScript Deferred, Delay JavaScript Execution with no exclusions.
5. Visit any page on the site and links will not be clickable.
6. Add `consent.cookiebot.com` to exclude from Delay JS to resolve the issue.

Other options don't cause issues on my and customer's sites, the script is not being processed by Minify JS option.

Ticket: https://secure.helpscout.net/conversation/1930304093/350485/

**Comment by NataliaDrause** [comment]
Another case where Delaying `consent.cookiebot.com` is blocking all the links.

Ticket: https://secure.helpscout.net/conversation/1956525959/357496/

**Comment by viobru** [comment]
Another case where the links were blocked by Dejay JS.

Ticket: https://secure.helpscout.net/conversation/2038330415/375039

**Comment by piotrbak** [comment]
Excluded `consent.cookiebot.com/uc.js` from Delay JS

**Comment by NataliaDrause** [comment]
Site in [this ticket](https://secure.helpscout.net/conversation/2066140207/381232/) is affected by the fix.

Performance drops significantly when cookiebot script is excluded.

Removing the exclusion brings good scores back and nothing is broken.
